system-contract-dapp-playground/src/api/hedera/tokenCreateCustom-interactions/index.ts

```diff
@@ -0,0 +1,237 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import { prepareHederaTokenKeyArray } from '@/utils/contract-interactions/HTS/helpers';
+import { Contract, isAddress } from 'ethers';
+
+/**
+ * @dev creates a Hedera fungible token
+ *
+ * @dev integrates tokenCreateCustomContract.createFungibleTokenPublic() and tokenCreateCustomContract.createFungibleTokenWithCustomFeesPublic()
+ *
+ * @param baseContract: ethers.Contract
+ *
+ * @param name: string
+ *
+ * @param symbol: string
+ *
+ * @param memo: string
+ *
+ * @param initialTotalSupply: number
+ *
+ * @param maxSupply: number
+ *
+ * @param decimals: number
+ *
+ * @param freezeDefaultStatus: boolean
+ *
+ * @param treasury: string
+ *
+ * @param inputKeys: CommonKeyObject[],
+ *
+ * @return Promise<TokenCreateCustomSmartContractResult>
+ *
+ * @see https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L136
+ *      for more information on the purposes of the params
+ */
+export const createHederaFungibleToken = async (
+  baseContract: Contract,
+  name: string,
+  symbol: string,
+  memo: string,
+  initialTotalSupply: number,
+  maxSupply: number,
+  decimals: number,
+  freezeDefaultStatus: boolean,
+  treasury: string,
+  inputKeys: CommonKeyObject[],
+  feeTokenAddress?: string
+): Promise<TokenCreateCustomSmartContractResult> => {
+  // sanitize params
+  if (initialTotalSupply < 0) {
+    return { err: 'initial total supply cannot be negative' };
+  } else if (maxSupply < 0) {
+    return { err: 'max supply cannot be negative' };
+  } else if (decimals < 0) {
+    return { err: 'decimals cannot be negative' };
+  } else if (!isAddress(treasury)) {
+    return { err: 'invalid treasury address' };
+  } else if (feeTokenAddress && !isAddress(feeTokenAddress)) {
+    return { err: 'invalid fee token address' };
+  }
+
+  // prepare keys array
+  const keyRes = prepareHederaTokenKeyArray(inputKeys);
+
+  // handle error
+  if (keyRes.err) {
+    return { err: keyRes.err };
+  }
+
+  try {
+    let tokenCreateTx;
+    if (feeTokenAddress) {
+      tokenCreateTx = await baseContract.createFungibleTokenWithCustomFeesPublic(
+        treasury,
+        feeTokenAddress,
+        name,
+        symbol,
+        memo,
+        initialTotalSupply,
+        maxSupply,
+        decimals,
+        keyRes.hederaTokenKeys,
+        {
+          value: '35000000000000000000',
+          gasLimit: 1_000_000,
```

Comment:
Q: should these be parameter(s) for easy configurability?

Reply:
Oh great catch I got this 35 hbar number after a few tests for quite some time ago and never remember to got back to it. And yes you're right these should be parameters so can be passed in dynamically by the users. The fact is I didn't know we have a whole fee calculator on the Hedera website, so people can use the calculator to determine what the fee they should pass in. But yeah pushing the fix soon

Reply:
Hmm on a second thought, it seems like a solid idea to have 'msgValue' set as a dynamic parameter. As for the 1_000_000 gas limit, I'm inclined to keep it as is for a smoother UX. Regarding 'msgValue' on the front-end, we can provide a link to the calculator, allowing users to determine the fee. However, adjusting the gas limit might confuse users, and I remember couple of time my transaction got rejected if the gas limits is improperly set. Any unused gas eventually gets returned to users, so there's no real loss on their end. Let me know what you think

Reply:
Yeah make `msg.value` an input
gasLimit refund max is 20% of what you put in so it's not all refunded.
The gasLimit changes per operation I would set it closer to what's needed.
DOn't need to hold up on the 1 million has limit but maybe open a ticket to revisit after phase 3 to have more realistic numbers or remove it so that `eth_estimateGas` can be used once HIP 584 is all done.

Reply:
Yessir! Thanks for the input!